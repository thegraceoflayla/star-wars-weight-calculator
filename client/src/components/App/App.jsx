import React from 'react';
import $ from 'jquery';
import Selector from '../Selector/Selector.jsx'

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      people: [],
      selectedPerson: 0,
      displayStartIndex: 0,
      weight: null,
      planet: '',
      gravity: 1
    }
    this.updatePerson = this.updatePerson.bind(this);
    this.rotateSelection = this.rotateSelection.bind(this);
  }

  updatePerson(e) {
    this.setState((state) => {
      state.selectedPerson = Number(e.target.id) + state.displayStartIndex;
      return state;
    }, () => this.findWeight(this.state.people[this.state.selectedPerson].homeworld));
  }

  rotateSelection(e) {
    this.setState((state) => {
      if (e.target.value === 'left') {
        state.displayStartIndex = Math.max(state.displayStartIndex - 5, 0);
      } else {
        state.displayStartIndex = Math.min(state.displayStartIndex + 5, state.people.length - 5);
      }
      return state;
    })
  }

  findWeight(planetLink) {
    $.ajax({
      url: planetLink,
      type: 'GET',
      success: ({name, gravity}) => {
        this.setState((state) => {
          state.planet = name;
          let parsedGravity = gravity.length > 5 ? Number(gravity.slice(0, gravity.indexOf(' '))) : Number(gravity);
          let parsedMass = Number(state.people[state.selectedPerson].mass.replace(/,/g, ''));
          state.gravity = gravity !== 'N/A' && gravity !== 'unknown' ? parsedGravity : -1;
          state.weight = Math.round(Math.abs(state.gravity) * parsedMass * 2.205);
          return state;
        });
      }
    })
  }

  populatePeople(data) {
    this.setState((state) => {
      state.people = state.people.concat(data.results
        .map(({name, mass, homeworld}) => ({name, mass, homeworld}))
        .filter(person => person.mass !== 'unknown' && person.homeworld !== 'unknown')
      );
      return state;
    });

    if (data.next !== null) {
      $.ajax({
        url: data.next,
        type: 'GET',
        success: (results) => this.populatePeople(results)
      })
    }
  }

  componentDidMount() {
    $.ajax({
      url: 'https://swapi.dev/api/people',
      type: 'GET',
      success: (results) => {
        this.populatePeople(results);
        this.findWeight(results.results[0].homeworld);
      }
    })
  }

  render() {
    let {people, selectedPerson, weight, planet, displayStartIndex, gravity} = this.state;
    let currentSelection = people.slice(displayStartIndex, displayStartIndex + 5);
    let displayText = gravity > -1 ?
      `${people[selectedPerson]?.name} weighs ${weight} pounds on their homeworld of ${planet}.` :
      `Hmm, we don't know what the gravity is like on ${people[selectedPerson]?.name}'s homeworld, but here on Earth, they'd weigh ${weight} pounds.`

    return people[0] ? (
      <div id="app" className="col m-5" style={{'backgroundColor': 'rgba(220,220,220,.6)', 'borderRadius': '2vw'}}>
        <h1 id="info" className="text-center row-8 m-4">{displayText}</h1>
        <Selector
          people={currentSelection}
          selectedPerson={people[selectedPerson].name}
          select={this.updatePerson}
          rotateDisplay={this.rotateSelection}
        />
      </div>
    ) : (
      <div id="loading">Loading...</div>
    )
  }
}

export default App;