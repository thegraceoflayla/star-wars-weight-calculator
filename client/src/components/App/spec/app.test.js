import React from 'react';
import { render, screen, act, waitFor, regeneratorRuntime, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import App from '../App.jsx';

it('shows loading message prior to initial API call', (done) => {
  const loading = renderer.create(<App />);
  expect(loading.toJSON().children[0]).toBe('Loading...');
  expect(loading).toMatchSnapshot();
  done();
});

it('renders after initial API call is complete', (done) => {
  const app = renderer.create(<App />);
  setTimeout(() => {
    expect(app.toJSON().children[0].children[0]).toBe('Luke Skywalker weighs 170 pounds on their homeworld of Tatooine.');
    expect(app).toMatchSnapshot();
    done();
  }, 2000);
});
