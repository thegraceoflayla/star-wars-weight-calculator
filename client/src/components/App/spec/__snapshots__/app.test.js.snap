// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders after initial API call is complete 1`] = `
<div
  className="col m-5"
  id="app"
  style={
    Object {
      "backgroundColor": "rgba(220,220,220,.6)",
      "borderRadius": "2vw",
    }
  }
>
  <h1
    className="text-center row-8 m-4"
    id="info"
  >
    Luke Skywalker weighs 170 pounds on their homeworld of Tatooine.
  </h1>
  <div
    className="col align-center m-4"
    data-testid="selector"
    id="selector"
  >
    <div
      className="row row-1 justify-content-center m-2"
      id="people"
    >
      <button
        className="btn btn-secondary btn-lg m-1"
        data-testid="person"
        id={0}
        onClick={[Function]}
      >
        Luke Skywalker
      </button>
      <button
        className="btn btn-outline-secondary btn-lg m-1"
        data-testid="person"
        id={1}
        onClick={[Function]}
      >
        C-3PO
      </button>
      <button
        className="btn btn-outline-secondary btn-lg m-1"
        data-testid="person"
        id={2}
        onClick={[Function]}
      >
        R2-D2
      </button>
      <button
        className="btn btn-outline-secondary btn-lg m-1"
        data-testid="person"
        id={3}
        onClick={[Function]}
      >
        Darth Vader
      </button>
      <button
        className="btn btn-outline-secondary btn-lg m-1"
        data-testid="person"
        id={4}
        onClick={[Function]}
      >
        Leia Organa
      </button>
    </div>
    <div
      className="row row-1 justify-content-center"
      id="scroll-buttons"
    >
      <button
        className="btn btn-primary m-2 col-1"
        id="left-button"
        onClick={[Function]}
        value="left"
      >
        <i
          className="fa-solid fa-chevron-left"
        />
      </button>
      <button
        className="btn btn-primary m-2 col-1"
        id="right-button"
        onClick={[Function]}
        value="right"
      >
        <i
          className="fa-solid fa-chevron-right"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`shows loading message prior to initial API call 1`] = `
<div
  id="loading"
>
  Loading...
</div>
`;
